File: app/src/main/java/shop/ineed/app/ineed/activity/DetailsProductsActivity.java
package shop.ineed.app.ineed.activity;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Build;
import android.support.design.widget.AppBarLayout;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v7.graphics.Palette;
import android.view.MenuItem;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import org.parceler.Parcels;

import shop.ineed.app.ineed.R;
import shop.ineed.app.ineed.domain.Product;
import shop.ineed.app.ineed.fragments.DetailsProductFragment;
import shop.ineed.app.ineed.util.Base64;

public class DetailsProductsActivity extends BaseActivity implements AppBarLayout.OnOffsetChangedListener {

    private String mImageProduct;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().requestFeature(Window.FEATURE_CONTENT_TRANSITIONS);
        setContentView(R.layout.activity_details_products);
        enableToolbar();

        Product product = Parcels.unwrap(getIntent().getParcelableExtra("product"));
        mImageProduct = product.getPictures().get(0);

        getSupportActionBar().setTitle(product.getName());
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        AppBarLayout appBarLayout = (AppBarLayout) findViewById(R.id.appbar);


            if (Build.VERSION.SDK_INT >= 21) {
                Window window = getWindow();
                window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
                appBarLayout.addOnOffsetChangedListener(this);
            }


        if (savedInstanceState == null) {
            DetailsProductFragment productFragment = new DetailsProductFragment();
            productFragment.setArguments(getIntent().getExtras());
            getSupportFragmentManager().beginTransaction().add(R.id.detailsProductFragment, productFragment).commit();
        }

        initViews();
    }

    private void initViews() {
        ImageView ivProduct = (ImageView) findViewById(R.id.ivProduct);
        Picasso.with(this).load(mImageProduct).into(ivProduct);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            String imageTransitionName = getIntent().getExtras().getString(ProductsActivity.EXTRA_PRODUCT_IMAGE_TRANSITION_NAME);
            ivProduct.setTransitionName(imageTransitionName);
        }
    }

    @Override
    public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
        int scrollRange = -1;
        //Initialize the size of the scroll
        if (scrollRange == -1) {
            scrollRange = appBarLayout.getTotalScrollRange();
        }
        //Check if the view is collapsed
        if (scrollRange + verticalOffset == 0) {
            getToolbar().setBackgroundColor(ActivityCompat.getColor(this, R.color.colorPrimary));
        } else {
            getToolbar().setBackgroundColor(Color.TRANSPARENT);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(item.getItemId() == android.R.id.home){
            finish();
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/shop/ineed/app/ineed/activity/DetailsProductsActivity.kt
package shop.ineed.app.ineed.activity

import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import com.google.firebase.crash.FirebaseCrash
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener

import kotlinx.android.synthetic.main.activity_details_products.*

import shop.ineed.app.ineed.R
import shop.ineed.app.ineed.domain.Product
import shop.ineed.app.ineed.domain.util.LibraryClass
import shop.ineed.app.ineed.fragments.DetailsProductFragment
import shop.ineed.app.ineed.util.ImageUtils

class DetailsProductsActivity : BaseActivity(){

    private var imageProduct: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_details_products)

        // Toolbar
        enableToolbar()

        val product = intent.getParcelableExtra<Product>("product")
        imageProduct = product.pictures[0]

        supportActionBar?.title = product.name
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val reference = LibraryClass.getFirebase().child("products").child(product.id)
        reference.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val product = dataSnapshot.getValue(Product::class.java)
                Log.i("STORE", product!!.name)

                if(savedInstanceState == null){
                    val productFragment = DetailsProductFragment()
                    val bundle = Bundle()
                    bundle.putParcelable("product", product)
                    productFragment.arguments = bundle
                    supportFragmentManager.beginTransaction().replace(R.id.detailsProductFragment, productFragment).commit()
                    Log.i("SAVED", "OK")
                }
            }
            override fun onCancelled(databaseError: DatabaseError) {
                FirebaseCrash.log(databaseError.message)
            }
        })

        // Loading
        swipeRefreshProduct.setColorSchemeResources(
                R.color.colorAccent,
                R.color.colorPrimary,
                R.color.colorPrimaryDark
        )
        swipeRefreshProduct.isRefreshing = true

        initViews()
    }

    fun disableToRefresh(){
        swipeRefreshProduct.isRefreshing = false
        swipeRefreshProduct.isEnabled = false
    }

    private fun initViews() {
        ImageUtils.displayImageFromUrl(this, imageProduct, ivProductDetailsProduct)
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }
}
